File: javascript/872.Leaf-Similar-Trees.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root1
 * @param {TreeNode} root2
 * @return {boolean}
 */
var leafSimilar = function (root1, root2) {
  const leafs1 = [];
  const leafs2 = [];

  const dfs = (node, leafs) => {
    if (!node) {
      return;
    }

    if (!node.left && !node.right) {
      leafs.push(node.val);
    }

    dfs(node.left, leafs);
    dfs(node.right, leafs);
  };

  dfs(root1, leafs1);
  dfs(root2, leafs2);

  if (leafs1.length !== leafs2.length) {
    return false;
  }

  for (let i = 0; i < leafs1.length; i++) {
    if (leafs1[i] !== leafs2[i]) {
      return false;
    }
  }

  return true;
};

File: python/872.Leaf-Similar-Trees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        leaf1 = []
        leaf2 = []

        def dfs(node, leaf):
            if not node:
                return
            if not node.left and not node.right:
                leaf.append(node.val)
            dfs(node.left, leaf)
            dfs(node.right, leaf)

        dfs(root1, leaf1)
        dfs(root2, leaf2)

        if len(leaf1) != len(leaf2):
            return False

        for i in range(len(leaf1)):
            if (leaf1[i] != leaf2[i]):
                return False

        return True